### ISSUE TITLE
Native support for categorical splits

### ISSUE THREAD

**Issue by RAMitchell** [rapidsai/legate-boost#29]
Existing GBM implementations present two approaches:

1) Implement customised categorical split logic in the tree construction. Sort gradients/hessians of all categories within a feature and find the optimal split. LightGBM/XGBoost follow this approach.

2) Preprocess the dataset to encode new features carrying category information. Catboost uses this approach. It is also implemented here: https://contrib.scikit-learn.org/category_encoders/catboost.html

1 is problematic in that it creates bias towards selecting categorical features over numeric features. Our implementation should ideally be unbiased. Perhaps there is a way to randomly select a single split, like we do for numeric features, in an unbiased way?

2 is problematic in that it potentially involves expensive distributed shuffling or data movement and also requires making a copy of the input matrix. 1) would be significantly less memory hungry.

**Comment by RAMitchell** [comment]
Am leaning right now towards implementing 2. as it has good performance in the literature and also generalises to all model types in legate-boost instead of just trees.

Implementing the tree version is also much more complicated and requires modifying cuda kernels and extending the storage format  for trees.

**Comment by RAMitchell** [comment]
#214 implemented an sklearn encoder. I will leave this open and implement a catboost style encoder in addition.
